tests/framework/microvms.py

```diff
@@ -0,0 +1,126 @@
+        self._vm = vm
+
+    @property
+    def config(self):
+        """Return machine config artifact."""
+        return self._config
+
+    @property
+    def kernel(self):
+        """Return the kernel artifact."""
+        return self._kernel
+
+    @property
+    def disks(self):
+        """Return an array of block file paths."""
+        return self._disks
+
+    @property
+    def ssh_key(self):
+        """Return ssh key artifact linked to the root block device."""
+        return self._ssh_key
+
+    @property
+    def vm(self):
+        """Return the Microvm object intsance."""
+        return self._vm
+
+
+# VM base class to abstract away the s3 artifacts we are using
+# to specify the configuration
+# Too few public methods (1/2) (too-few-public-methods)
+# pylint: disable=R0903
+class VMBase:
+    """Base class for constructing microvms."""
+
+    @classmethod
+    def from_artifacts(cls, bin_cloner_path, config,
+                       kernel, disks, cpu_template, start=False):
+        """Spawns a new Firecracker and applies specified config."""
+        artifacts = ArtifactCollection(_test_images_s3_bucket())
+        config = artifacts.microvms(keyword=config)[0]
+        kernel = artifacts.kernels(keyword=kernel)[0]
+        disks = artifacts.disks(keyword=disks)
+        config.download()
+        kernel.download()
+        attached_disks = []
+        for disk in disks:
+            disk.download()
+            attached_disks.append(disk.copy())
+
+        vm_builder = MicrovmBuilder(bin_cloner_path)
+
+        # SSH key is attached to root disk artifact.
+        # Builder will download ssh key in the VM root.
+        ssh_key = disks[0].ssh_key()
+
+        # Create a fresh microvm from aftifacts.
+        basevm = vm_builder.build(kernel=kernel,
+                                  disks=attached_disks,
+                                  ssh_key=ssh_key,
+                                  config=config,
+                                  cpu_template=cpu_template)
+
+        network_config = net_tools.UniqueIPv4Generator.instance()
+        # Host ip: 192.168.0.1
+        # Guest ip: 192.168.0.2
+        # TODO: we need to refactor the network config using artifacts.
+        basevm.ssh_network_config(network_config,
+                                  '1',
+                                  tapname="tap0")
+        if start:
+            basevm.start()
+            ssh_connection = net_tools.SSHConnection(basevm.ssh_config)
+
+            # Verify if we can run commands in guest via ssh.
+            exit_code, _, _ = ssh_connection.execute_command("sync")
+            assert exit_code == 0
+
+        return VmInstance(config, kernel, attached_disks, ssh_key, basevm)
+
+
+class C3nano(VMBase):
+    """Create VMs with 2vCPUs and 256 MB RAM."""
+
+    @classmethod
+    def spawn(cls, bin_cloner_path, start=False):
+        """Spawns and optionally starts the vm."""
+        return cls.from_artifacts(bin_cloner_path,
```

Comment:
We should call the constructor by specifying the class parameters explicitly, in order to make it easier to read.